packages/contracts/contracts/bridge/xERC20Hyperlane.sol

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.22;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IXERC20 } from "./interface/IXERC20.sol";
+import { IMailbox } from "./hyperlane/interfaces/IMailbox.sol";
+
+library TypeCasts {
+  // alignment preserving cast
+  function addressToBytes32(address _addr) internal pure returns (bytes32) {
+    return bytes32(uint256(uint160(_addr)));
+  }
+
+  // alignment preserving cast
+  function bytes32ToAddress(bytes32 _buf) internal pure returns (address) {
+    return address(uint160(uint256(_buf)));
+  }
+}
+
+contract xERC20Hyperlane is Ownable {
+  using TypeCasts for address;
+  using TypeCasts for bytes32;
+
+  uint256 public feeBps;
+  mapping(address => mapping(uint32 => address)) public mappedTokens;
+  mapping(uint32 => address) public mappedBridges;
+  IMailbox immutable mailbox;
+
+  event TokenMapped(address indexed _token, uint32 indexed _chainId, address indexed _dstToken);
+  event BridgeMapped(uint32 indexed _chainId, address indexed _bridge);
+  event FeeBpsSet(uint256 indexed _feeBps);
+  event TokenSent(
+    address indexed _token,
+    uint256 _amount,
+    address indexed _to,
+    uint32 indexed _dstChainId,
+    bytes32 _guid
+  );
+  event TokenReceived(
+    address indexed _token,
+    uint256 _amount,
+    address indexed _to,
+    uint32 indexed _srcChainId,
+    bytes32 _guid
+  );
+
+  error TokenNotSet();
+  error ChainIdNotSet();
+  error DestinationBridgeNotSet(uint32 _chainId);
+  error OriginNotAllowed(uint32 _chainId, address _sender);
+
+  /**
+   * @notice Only accept messages from an Hyperlane Mailbox contract
+   */
+  modifier onlyMailbox() {
+    require(msg.sender == address(mailbox), "MailboxClient: sender not mailbox");
+    _;
+  }
+
+  constructor(uint256 _feeBps, address _mailbox) Ownable() {
+    mailbox = IMailbox(_mailbox);
+    feeBps = _feeBps;
+  }
+
+  // ADMIN FUNCTIONS
+  function setFeeBps(uint256 _feeBps) public onlyOwner {
+    feeBps = _feeBps;
+    emit FeeBpsSet(_feeBps);
+  }
+
+  function setMappedToken(uint32 _chainId, address _srcToken, address _dstToken) public onlyOwner {
```

Comment:
Hyperlane normally used the notion of domains like destinationDomain instead of chainId since there have been cases of duplicate chainId or setting an arbitrary domain for non-EVM domains